lib/stages/hdf5FileWrite.cpp

```diff
@@ -0,0 +1,240 @@
+#include "StageFactory.hpp"    // for REGISTER_KOTEKAN_STAGE, StageMakerTemplate
+#include "buffer.hpp"          // for Buffer, get_metadata_container, mark_frame_empty, regis...
+#include "bufferContainer.hpp" // for bufferContainer
+#include "chimeMetadata.hpp"   // for chimeMetadata
+#include "chordMetadata.hpp"   // for chordMetadata
+#include "errors.h"
+#include "kotekanLogging.hpp"    // for ERROR, INFO
+#include "metadata.h"            // for metadataContainer
+#include "prometheusMetrics.hpp" // for Metrics, Gauge
+#include "visUtil.hpp"           // for current_time
+
+#undef NDEBUG
+
+#include <atomic>     // for atomic_bool
+#include <cassert>    // for assert
+#include <errno.h>    // for errno
+#include <exception>  // for exception
+#include <fcntl.h>    // for open, O_CREAT, O_WRONLY
+#include <functional> // for _Bind_helper<>::type, bind, function
+#include <hdf5.h>
+#include <iomanip>   // for setfill, setw
+#include <regex>     // for match_results<>::_Base_type
+#include <signal.h>  // for raise, SIGHUP
+#include <sstream>   // for ostringstream
+#include <stdexcept> // for runtime_error
+#include <stdint.h>  // for uint32_t, int32_t, uint8_t
+#include <stdio.h>   // for snprintf
+#include <stdlib.h>  // for exit
+#include <string>    // for string
+#include <unistd.h>  // for gethostname
+#include <vector>    // for vector
+
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+using kotekan::prometheus::Metrics;
+
+REGISTER_KOTEKAN_STAGE(hdf5FileWrite);
+
+std::atomic<uint32_t> hdf5FileWrite::n_finished{0};
+
+hdf5FileWrite::hdf5FileWrite(Config& config, const std::string& unique_name,
+                             bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&hdf5FileWrite::main_thread, this)) {
+
+    buf = get_buffer("in_buf");
+    register_consumer(buf, unique_name.c_str());
+    _base_dir = config.get<std::string>(unique_name, "base_dir");
+    _file_name = config.get<std::string>(unique_name, "file_name");
+    _prefix_hostname = config.get_default<bool>(unique_name, "prefix_hostname", true);
+    _exit_after_n_frames = config.get_default<uint32_t>(unique_name, "exit_after_n_frames", 0);
+    _exit_with_n_writers = config.get_default<uint32_t>(unique_name, "exit_with_n_writers", 0);
+}
+
+hdf5FileWrite::~hdf5FileWrite() {}
+
+void hdf5FileWrite::main_thread() {
+    std::string full_path;
+    hid_t fd = -1;
+
+    uint32_t frame_id = 0;
+    uint32_t frame_ctr = 0;
+    uint8_t* frame = nullptr;
+
+    auto& write_time_metric =
+        Metrics::instance().add_gauge("kotekan_hdf5filewrite_write_time_seconds", unique_name);
+    while (!stop_thread) {
+
+        // This call is blocking.
+        frame = wait_for_full_frame(buf, unique_name.c_str(), frame_id);
+        if (frame == nullptr)
+            break;
+
+        // Start timing the write time
+        const double st = current_time();
+
+        if (fd < 0) {
```

Comment:
Could check for H5I_UNINIT instead of < 0 here.

Reply:
Done.